// @seleniumhq/side-runtime/dist/index.js
"use strict";
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecordingSyncronizers = exports.CommandType = exports.WebDriverExecutor = exports.Variables = exports.Playback = void 0;
var playback_1 = require("./playback");
Object.defineProperty(exports, "Playback", { enumerable: true, get: function () { return __importDefault(playback_1).default; } });
var variables_1 = require("./variables");
Object.defineProperty(exports, "Variables", { enumerable: true, get: function () { return __importDefault(variables_1).default; } });
var webdriver_1 = require("./webdriver");
Object.defineProperty(exports, "WebDriverExecutor", { enumerable: true, get: function () { return __importDefault(webdriver_1).default; } });
const recording_syncronizer_1 = __importDefault(require("./recording-syncronizer"));
const recording_syncronizer_webdriver_1 = __importDefault(require("./recording-syncronizer-webdriver"));
__exportStar(require("./playback"), exports);
__exportStar(require("./plugins"), exports);
__exportStar(require("./types"), exports);
var command_node_1 = require("./playback-tree/command-node");
Object.defineProperty(exports, "CommandType", { enumerable: true, get: function () { return command_node_1.CommandType; } });
exports.RecordingSyncronizers = {
    createRecorderSyncronizer: recording_syncronizer_1.default,
    createRecorderSyncronizerForWebdriverExecutor: recording_syncronizer_webdriver_1.default,
};
